apple/swift-docc
# Support `deviceFrame:` parameter in `@Image` and `@Video` directives

Bug/issue #, if applicable: rdar://102183073

## Summary

When developing UI libraries or apps,
it can be useful to show previews of how things look
when rendered on a device, like a phone or a tablet.
This new experimental feature is designed to allow authors to wrap
images and videos in a frame, making it easier to create complex assets
and maintain them overtime.

Both `@Image` and `@Video` gain a `deviceFrame` parameter that
associates them with a device frame defined in the DocC catalog’s
`theme-settings.json` file.

This change is described on the Swift forums here:
https://forums.swift.org/t/support-for-device-frames/62151

## Dependencies

DocC-Render PR TBD

## Testing

### Test that `deviceFrame` works as expected

1. Add an `@Image` or `@Video` directive to a DocC article:

    ```md
    # My Great Article
    
    This is a great one!
    
    With a lot of great content.
    
    @Image(source: "sloth", deviceFrame: phone)
    ```

2. Confirm that the emitted RenderJSON references the `"phone"` device frame

### Test that `deviceFrame` is excluded from documentation

This is an experimental feature that shouldn't be advertised in documentation yet.

1. `swift run generate-symbol-graph`
2. `bin/preview-docs DocC`
3. <http://localhost:8080/documentation/docc/image>
4. Confirm that `@Image` and `@Video` declarations don't include `deviceFrame`:

    <img width="837" alt="Screenshot 2023-01-11 at 1 57 54 PM" src="https://user-images.githubusercontent.com/6750147/211926582-8dab11d3-b70d-4628-8d22-3fe40972e1c2.png">



## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.
